docs/RFCS/leader_evaluated_raft.md

```diff
@@ -0,0 +1,425 @@
+  // explicit method to finalize the writes if that is more convenient].
+  GetWrites() (Writes, error)
+}
+```
+
+With this strategy, the MVCC changes are fairly locally scoped and not too
+invasive.
+
+Naively, this can be achieved by wrapping a "regular" batch with an appropriate
+in-memory map which is populated on writes.
+
+However, using [RocksDB's `WriteBatch`
+format](https://github.com/facebook/rocksdb/blob/f38540b12ada4fe06598c42d4c084f9d920289ff/db/write_batch.cc#L10)
+seems opportune:
+
+```
+// WriteBatch::rep_ :=
+//    sequence: fixed64
+//    count: fixed32
+//    data: record[count]
+// record :=
+//    kTypeValue varstring varstring
+//    kTypeDeletion varstring
+//    kTypeSingleDeletion varstring
+//    kTypeMerge varstring varstring
+//    kTypeColumnFamilyValue varint32 varstring varstring
+//    kTypeColumnFamilyDeletion varint32 varstring varstring
+//    kTypeColumnFamilySingleDeletion varint32 varstring varstring
+//    kTypeColumnFamilyMerge varint32 varstring varstring
+// varstring :=
+//    len: varint32
+//    data: uint8[len]
+```
+
+* The format is straightforward to implement; the writes need to be serialized
+  anyway for the Raft proposal, and this appears to be an efficient format to
+  use, even when the underlying storage engine isn't based on RocksDB.
+* We currently use RocksDB and this is already the internal format used by a
+  RocksDB batch. We can completely avoid an additional copy and serialization
+  step in this case (i.e. construct the batch, borrow out the contained
+  representation to Raft until the command commits, and apply the batch).
+  Followers can create a `WriteBatch` directly from this representation.
+  Serialization is a key point in performance[1].
+* Again for RocksDB, we might also be able to piggy-back on existing
+  implementation to operate on the superposition of multiple `WriteBatch`es
+  (which would be needed to remove write-write blocking).
+
+The format should be relatively stable (likely only additions), but there might
+be future migrations coming up as RocksDB changes their underlying assumptions.
+We should consider checking in with them about this idea.
+
+Using `WriteBatch` also means that the writes are completely opaque to the
+stack (at least unless unserialized early, and loaded, and then accessed
+through MVCC operations), so the followers essentially can't act based on
+them. That should not be a problem (but could if we want the side effects
+to use information from the write set in a performant way).
+
+## Raft command application
+
+This will be a new, straightforward facility which applies a `RaftCmd` against
+the underlying `Engine` in a batch. The point of this whole change is that this
+part of the stack is very straightforward and with no side effects.
+
+`(*Replica).executeCmd` is likely a good place where most of this code would
+live.
+
+## Replay protection
+
+Replay protection is naively an issue because applying a logicless batch of
```

Comment:
Let's plan an audit of the code to remove anything which I added in the big transaction semantics refactoring to notice Raft replays. I just want to make sure we don't have code to identify Raft replays living in there once we move this stuff in front of Raft.